File: examples/python/x_api.py
"""X Chat API access via the X API SDK (XDK).

The bot's I/O layer: every REST call goes through the XDK Python client so
request signing, models, and endpoint paths stay in sync with the X API.

Authentication is an OAuth2 user access token (scopes ``dm.read`` + ``dm.write``)
read from the environment.
"""

from __future__ import annotations

import base64
import json
import urllib.error
import urllib.parse
import urllib.request
from typing import Any

from xdk import Client
from xdk.chat.models import SendMessageRequest

BASE_URL = "https://api.x.com"


class XChatClient:
    def __init__(self, access_token: str) -> None:
        # The XDK client picks OAuth2 user-context auth from the access token.
        self.client = Client(access_token=access_token)
        self._token = access_token

    def _post(self, path: str, body: dict[str, Any]) -> dict[str, Any]:
        """POST a chat endpoint not yet covered by the installed XDK release.

        Chat auth is a plain OAuth2 bearer token, so a direct call is
        equivalent to what the XDK sends.
        """
        req = urllib.request.Request(
            BASE_URL + path,
            data=json.dumps(body).encode(),
            headers={
                "Authorization": f"Bearer {self._token}",
                "Content-Type": "application/json",
            },
            method="POST",
        )
        with urllib.request.urlopen(req, timeout=30) as resp:
            return json.loads(resp.read().decode() or "{}")

    # -- Identity -----------------------------------------------------------

    def get_my_user_id(self) -> str:
        return str(self.client.users.get_me().data.id)

    def get_public_keys(self, user_id: str) -> list[dict[str, Any]]:
        """Fetch a user's registered public keys (for ECIES + verify, or to
        check your own before registering).

        Every field of the public_key resource (``public_key``,
        ``signing_public_key``, ``identity_public_key_signature``,
        ``public_key_version``, ``juicebox_config``) is always included; the
        route takes no ``public_key.fields`` parameter.
        """
        resp = self.client.users.get_public_key(user_id)
        data = resp.data or []
        items = data if isinstance(data, list) else [data]
        return [d.model_dump() if hasattr(d, "model_dump") else dict(d) for d in items]

    class RateLimited(Exception):
        """Raised when the public-key write bucket is exhausted (HTTP 429).

        The endpoint allows only a few writes per 24h; ``reset_epoch`` is when
        the window frees up. Retrying before then just fails again.
        """

        def __init__(self, reset_epoch: int | None) -> None:
            self.reset_epoch = reset_epoch
            super().__init__("public-key registration rate limited (HTTP 429)")

    def add_user_public_key(self, user_id: str, payload: dict[str, Any]) -> dict[str, Any]:
        """Register a public key: POST /2/users/{id}/public_keys.

        ``payload`` is the registration object from ``generate_keypairs`` in its
        snake_case wire form (``public_key`` object, ``version``,
        ``generate_version``). Raises ``RateLimited`` on 429 so the caller can
        stop instead of burning the daily budget.
        """
        path = f"/2/users/{user_id}/public_keys"
        req = urllib.request.Request(
            BASE_URL + path,
            data=json.dumps(payload).encode(),
            headers={
                "Authorization": f"Bearer {self._token}",
                "Content-Type": "application/json",
            },
            method="POST",
        )
        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                return json.loads(resp.read().decode() or "{}")
        except urllib.error.HTTPError as err:
            if err.code == 429:
                reset = err.headers.get("x-user-limit-24hour-reset")
                raise self.RateLimited(int(reset) if reset else None) from err
            raise

    def get_juicebox_config(self, user_id: str) -> tuple[str, str]:
        """Build the Juicebox config JSON + latest key version for ``ChatCore.unlock``.

        Only needed for the optional Juicebox (production) key-storage path.
        Every public_key field (``juicebox_config`` included) is always
        returned; the route takes no ``public_key.fields`` parameter.
        """
        resp = self.client.users.get_public_key(user_id)
        data = resp.data or []
        items = [
            d.model_dump() if hasattr(d, "model_dump") else dict(d)
            for d in (data if isinstance(data, list) else [data])
        ]
        latest = max(items, key=lambda d: int(d.get("public_key_version") or 0))
        # The X API `juicebox_config` object is accepted as-is: the core reads
        # `key_store_token_map_json` verbatim and auth tokens from `token_map`.
        return json.dumps(latest["juicebox_config"]), str(latest["public_key_version"])

    # -- Conversation events ------------------------------------------------

    def get_events(
        self,
        conversation_id: str,
        *,
        max_results: int = 50,
        pagination_token: str | None = None,
    ) -> dict[str, Any]:
        """GET the raw (encrypted) events for a conversation via the XDK.

        Each ``data`` item carries an ``encoded_event`` base64 blob the crypto
        core decrypts via ``decrypt_batch`` / ``decrypt_one``.
        """
        resp = self.client.chat.get_conversation_events(
            conversation_id.replace(":", "-"),
            max_results=max_results,
            pagination_token=pagination_token,
        )
        return resp.model_dump() if hasattr(resp, "model_dump") else dict(resp)

    # -- Conversation / key management ---------------------------------------

    def add_conversation_keys(self, conversation_id: str, body: dict[str, Any]) -> dict[str, Any]:
        """POST a prepared conversation-key change (initialize or rotate).

        ``body`` is the REST shape built by ``chat_core.prep_to_request``.
        For a 1:1, ``conversation_id`` may be the recipient's user ID; the
        server derives (and returns) the canonical conversation ID.
        """
        path = f"/2/chat/conversations/{conversation_id.replace(':', '-')}/keys"
        return self._post(path, body)

    def initialize_group(self) -> str:
        """Mint a new group conversation id (``g…``)."""
        out = self._post("/2/chat/conversations/group/initialize", {})
        return str((out.get("data") or {}).get("conversation_id") or "")

    def create_conversation(self, body: dict[str, Any]) -> dict[str, Any]:
        """POST /2/chat/conversations/group — create a group conversation.

        ``body`` carries ``conversation_id``, ``group_members``,
        ``group_admins``, and the two-signature key change from
        ``ChatCore.prepare_group_create``.
        """
        return self._post("/2/chat/conversations/group", body)

    def add_group_members(self, conversation_id: str, body: dict[str, Any]) -> dict[str, Any]:
        """POST /2/chat/conversations/{id}/members — add members to a group.

        ``body`` carries ``user_ids`` plus the rotated key change from
        ``ChatCore.prepare_group_members_change``.
        """
        path = f"/2/chat/conversations/{conversation_id}/members"
        return self._post(path, body)

    # -- Media (encrypted blobs) ---------------------------------------------

    UPLOAD_CHUNK = 3 * 1024 * 1024

    def upload_media(self, conversation_id: str, ciphertext: bytes) -> str:
        """Upload an encrypted media blob; returns its ``media_hash_key``.

        Three-step flow: initialize (returns an upload session and the hash
        key), append (3 MB segments), finalize. The media endpoints take the
        colon form of the conversation id in the body.
        """
        conv = conversation_id.replace("-", ":")
        init = self._post(
            "/2/chat/media/upload/initialize",
            {"conversation_id": conv, "total_bytes": len(ciphertext)},
        )
        data = init.get("data") or {}
        session_id = data.get("session_id") or data.get("sessionId")
        media_hash_key = data.get("media_hash_key") or data.get("mediaHashKey")
        if not session_id or not media_hash_key:
            raise RuntimeError(f"media upload initialize failed: {init}")

        segment = 0
        for offset in range(0, len(ciphertext), self.UPLOAD_CHUNK):
            chunk = ciphertext[offset : offset + self.UPLOAD_CHUNK]
            self._post(
                f"/2/chat/media/upload/{session_id}/append",
                {
                    "conversation_id": conv,
                    "media_hash_key": media_hash_key,
                    "segment_index": str(segment),
                    "media": base64.b64encode(chunk).decode(),
                },
            )
            segment += 1

        self._post(
            f"/2/chat/media/upload/{session_id}/finalize",
            {
                "conversation_id": conv,
                "media_hash_key": media_hash_key,
                "num_parts": str(segment),
            },
        )
        return str(media_hash_key)

    def download_media(self, conversation_id: str, media_hash_key: str) -> bytes:
        """Download an encrypted media blob as raw bytes.

        The response body is binary ciphertext — it must be read as bytes;
        any text decoding would corrupt it. The download path takes the
        hyphen form of the conversation id.
        """
        conv = urllib.parse.quote(conversation_id.replace(":", "-"), safe="")
        hash_key = urllib.parse.quote(media_hash_key, safe="")
        req = urllib.request.Request(
            f"{BASE_URL}/2/chat/media/{conv}/{hash_key}",
            headers={"Authorization": f"Bearer {self._token}"},
        )
        with urllib.request.urlopen(req, timeout=60) as resp:
            return resp.read()

    # -- Sending ------------------------------------------------------------

    def send_message(self, conversation_id: str, body: dict[str, str]) -> Any:
        """POST an encrypted message produced by ``ChatCore.encrypt_reply``.

        For a 1:1 conversation, ``conversation_id`` is the recipient's user ID;
        the server constructs the canonical conversation ID.
        """
        request = SendMessageRequest.model_validate(body)
        return self.client.chat.send_message(conversation_id.replace(":", "-"), request)

File: examples/js/src/x-api.mjs
/**
 * X Chat API access via the X API SDK (XDK).
 *
 * Every REST call goes through the XDK client so request signing, models, and
 * endpoint paths stay in sync with the X API. Authentication is an OAuth2 user
 * access token (scopes dm.read + dm.write).
 */
import { Client } from "@xdevplatform/xdk";

const BASE_URL = "https://api.x.com";

export class XChatClient {
  constructor(accessToken) {
    this.client = new Client({ accessToken });
    this.accessToken = accessToken;
  }

  async getMyUserId() {
    const resp = await this.client.users.getMe();
    return String(resp.data.id);
  }

  /**
   * Fetch a user's registered public keys (for ECIES + verification, or to
   * check your own before registering).
   *
   * Every field of the public_key resource (public_key, signing_public_key,
   * identity_public_key_signature, public_key_version, juicebox_config) is
   * always included; the route takes no public_key.fields parameter.
   */
  async getPublicKeys(userId) {
    const resp = await this.client.users.getPublicKey(userId);
    const data = resp.data ?? [];
    return Array.isArray(data) ? data : [data];
  }

  /**
   * Raised when the public-key write bucket is exhausted (HTTP 429).
   *
   * The endpoint allows only a few writes per 24h; `resetEpoch` is when the
   * window frees up. Retrying before then just fails again.
   */
  static RateLimited = class RateLimited extends Error {
    constructor(resetEpoch) {
      super("public-key registration rate limited (HTTP 429)");
      this.name = "RateLimited";
      this.resetEpoch = resetEpoch;
    }
  };

  /**
   * Register a public key: POST /2/users/{id}/public_keys.
   *
   * `payload` is the registration object from `generateKeypairs` in its
   * snake_case wire form (`public_key` object, `version`, `generate_version`).
   * Sent with a direct fetch — chat auth is a plain OAuth2 bearer token, so
   * this is equivalent to what the XDK would send. Throws `RateLimited` on 429
   * so the caller can stop instead of burning the strict daily budget.
   */
  async addUserPublicKey(userId, payload) {
    const resp = await fetch(`${BASE_URL}/2/users/${encodeURIComponent(userId)}/public_keys`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.accessToken}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(payload),
    });
    if (resp.status === 429) {
      const reset = resp.headers.get("x-user-limit-24hour-reset");
      throw new XChatClient.RateLimited(reset ? Number(reset) : null);
    }
    if (!resp.ok) {
      throw new Error(`add public key failed: HTTP ${resp.status} ${await resp.text()}`);
    }
    const text = await resp.text();
    return text ? JSON.parse(text) : {};
  }

  /**
   * Build the Juicebox config JSON + latest key version for `unlock`/`setup`.
   *
   * The browser-safe WASM binding has no key export, so the registration
   * example persists the identity in Juicebox rather than a local blob; this
   * fetches the `juicebox_config` object the SDK needs to reach the realms.
   */
  async getJuiceboxConfig(userId) {
    const resp = await this.client.users.getPublicKey(userId);
    const data = resp.data ?? [];
    const items = Array.isArray(data) ? data : [data];
    const versionOf = (d) => Number(d.publicKeyVersion ?? d.public_key_version ?? 0);
    const latest = items.reduce(
      (best, d) => (versionOf(d) >= versionOf(best) ? d : best),
      items[0] ?? {},
    );
    const config = latest.juiceboxConfig ?? latest.juicebox_config;
    if (!config) {
      throw new Error(
        "no juicebox_config on the account yet. It is created by " +
          "POST /2/users/:id/public_keys, so register a public key first — " +
          "src/register.mjs does this: createChat with no config, " +
          "generateKeypairs, POST, then updateConfig + setup(pin). Once " +
          "provisioned, juicebox_config is always returned.",
      );
    }
    // Passed to the SDK as-is: it reads `key_store_token_map_json` verbatim.
    return {
      configJson: JSON.stringify(config),
      version: String(latest.publicKeyVersion ?? latest.public_key_version ?? "1"),
    };
  }

  /** GET the raw (encrypted) events for a conversation. */
  async getEvents(conversationId, { maxResults = 50, paginationToken } = {}) {
    return this.client.chat.getConversationEvents(conversationId.replaceAll(":", "-"), {
      maxResults,
      paginationToken,
    });
  }

  /** POST an encrypted message produced by ChatCore.encryptReply. */
  async sendMessage(conversationId, body) {
    return this.client.chat.sendMessage(conversationId.replaceAll(":", "-"), {
      messageId: body.message_id,
      encodedMessageCreateEvent: body.encoded_message_create_event,
      encodedMessageEventSignature: body.encoded_message_event_signature,
      conversationToken: body.conversation_token,
    });
  }

  // -- Conversation / key management ----------------------------------------

  /**
   * POST a prepared conversation-key change (initialize or rotate).
   * `body` is the request shape built by `prepToRequest`. For a 1:1,
   * `conversationId` may be the recipient's user ID; the server derives
   * (and returns) the canonical conversation ID.
   */
  async addConversationKeys(conversationId, body) {
    return this.client.chat.initializeConversationKeys(
      conversationId.replaceAll(":", "-"),
      body,
    );
  }

  /** Mint a new group conversation id (`g…`). */
  async initializeGroup() {
    const resp = await this.client.chat.initializeGroup();
    return String(resp.data?.conversationId ?? "");
  }

  /**
   * Create a group conversation. `body` carries `conversationId`,
   * `groupMembers`, `groupAdmins`, and the two-signature key change from
   * `ChatCore.prepareGroupCreate`.
   */
  async createConversation(body) {
    return this.client.chat.createConversation(body);
  }

  /**
   * Add members to a group. `body` carries `userIds` plus the rotated key
   * change from `ChatCore.prepareGroupMembersChange`.
   */
  async addGroupMembers(conversationId, body) {
    return this.client.chat.addGroupMembers(conversationId, body);
  }

  // -- Media (encrypted blobs) ------------------------------------------------

  static UPLOAD_CHUNK = 3 * 1024 * 1024;

  /**
   * Upload an encrypted media blob; returns its `media_hash_key`.
   *
   * Three-step flow: initialize (returns an upload session and the hash
   * key), append (3 MB segments), finalize. The media endpoints take the
   * colon form of the conversation id in the body.
   */
  async uploadMedia(conversationId, ciphertext) {
    const conv = conversationId.replaceAll("-", ":");
    const init = await this.client.chat.mediaUploadInitialize({
      conversationId: conv,
      totalBytes: ciphertext.length,
    });
    const sessionId = init.data?.sessionId ?? init.data?.session_id;
    const mediaHashKey = init.data?.mediaHashKey ?? init.data?.media_hash_key;
    if (!sessionId || !mediaHashKey) {
      throw new Error(`media upload initialize failed: ${JSON.stringify(init)}`);
    }

    let segment = 0;
    for (let offset = 0; offset < ciphertext.length; offset += XChatClient.UPLOAD_CHUNK) {
      const chunk = ciphertext.subarray(offset, offset + XChatClient.UPLOAD_CHUNK);
      await this.client.chat.mediaUploadAppend(sessionId, {
        conversationId: conv,
        mediaHashKey,
        segmentIndex: String(segment),
        media: Buffer.from(chunk).toString("base64"),
      });
      segment++;
    }

    await this.client.chat.mediaUploadFinalize(sessionId, {
      conversationId: conv,
      mediaHashKey,
      numParts: String(segment),
    });
    return String(mediaHashKey);
  }

  /**
   * Download an encrypted media blob as raw bytes.
   *
   * Uses fetch directly: the body is binary ciphertext and must be read as
   * an ArrayBuffer — any text decoding corrupts bytes above 0x7f. The
   * download path takes the hyphen form of the conversation id.
   */
  async downloadMedia(conversationId, mediaHashKey) {
    const conv = encodeURIComponent(conversationId.replaceAll(":", "-"));
    const hashKey = encodeURIComponent(mediaHashKey);
    const resp = await fetch(`${BASE_URL}/2/chat/media/${conv}/${hashKey}`, {
      headers: { Authorization: `Bearer ${this.accessToken}` },
    });
    if (!resp.ok) {
      throw new Error(`media download failed: HTTP ${resp.status}`);
    }
    return new Uint8Array(await resp.arrayBuffer());
  }
}
